python/tvm/contrib/popen_pool.py

```diff
@@ -213,12 +223,19 @@ def send(self, fn, args=(), kwargs=None, timeout=None):
         # pylint: disable=import-outside-toplevel
         import cloudpickle
 
+        if self._proc is not None and self._maximum_uses and self._remaining_uses == 0:
```

Comment:
we can move this block below lines 248-250 and remove self._proc check. It will also immediately free resources after last remaining use instead of needing additional send call.

Reply:
We can't quite do that because we have to wait for the `recv` call (at the start of `send` is "ok" only by relying on the contract above):

```
        The caller must call recv before calling the next send in
        order to make sure the timeout and child process exit
        won't affect the later requests.
```

But, I can move this to the end of `recv` if you'd like? I'm indifferent -- I put it in `send` to try consolidate (most) of the logic in one function but I certainly don't feel strongly about it.

Reply:
Ah I see, nah let's keep it consolidated